mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -164,21 +181,51 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
+            .next();
+    }
+
+    private Mono<ReIndexingEntry> createReindexingEntryFromFailure(ReIndexingFailure failure) {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXER_PERFORMER_USER);
+
+        return mailboxSessionMapperFactory.getMailboxMapper(mailboxSession)
+            .findMailboxByIdReactive(failure.getMailboxId())
+            .flatMap(mailbox -> fullyReadMessage(mailboxSession, mailbox, failure.getUid())
+                .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message)));
+    }
+
+    private Flux<ReIndexingEntry> findMessagesInMailbox(Mailbox mailbox, MailboxSession mailboxSession) {
+        MessageMapper messageMapper = mailboxSessionMapperFactory.getMessageMapper(mailboxSession);
+
+        return messageSearchIndex.deleteAll(mailboxSession, mailbox.getMailboxId())
+            .thenMany(messageMapper.listAllMessageUids(mailbox))
+            .flatMap(uid -> messageMapper.findInMailboxReactive(mailbox, MessageRange.one(uid), MessageMapper.FetchType.Full, UNLIMITED))
+            .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message));
+    }
+
+    private Mono<Task.Result> reIndexMessages(Flux<ReIndexingEntry> entries, RunningOptions runningOptions, ReprocessingContext reprocessingContext) {
+        return throttle(entries, Duration.ofSeconds(1), runningOptions.getMessagesPerSecond())
+            .flatMap(entry -> reIndexMessage(entry.getMailboxSession(), entry.getMailbox(), reprocessingContext, entry.getMessage()))
+            .reduce(Task::combine)
+            .switchIfEmpty(Mono.just(Result.COMPLETED));
+    }
+
+    private Mono<Task.Result> reIndexMessage(MailboxSession mailboxSession, Mailbox mailbox, ReprocessingContext reprocessingContext, MailboxMessage message) {
+        return Mono.fromCallable(() -> messageSearchIndex.add(mailboxSession, mailbox, message))
             .doOnNext(any -> reprocessingContext.recordSuccess())
+            .thenReturn(Result.COMPLETED)
             .onErrorResume(e -> {
-                LOGGER.warn("ReIndexing failed for {} {}", mailbox.generateAssociatedPath(), uid, e);
-                reprocessingContext.recordFailureDetailsForMessage(mailbox.getMailboxId(), uid);
+                LOGGER.warn("ReIndexing failed for {} {}", mailbox.generateAssociatedPath(), message.getUid(), e);
+                reprocessingContext.recordFailureDetailsForMessage(mailbox.getMailboxId(), message.getUid());
                 return Mono.just(Result.PARTIAL);
             });
     }
 
-    private Mono<MailboxMessage> fullyReadMessage(MailboxSession mailboxSession, Mailbox mailbox, MessageUid mUid) {
-        return mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
-            .findInMailboxReactive(mailbox, MessageRange.one(mUid), MessageMapper.FetchType.Full, SINGLE_MESSAGE)
-            .next();
+    private <V> Flux<V> throttle(Flux<V> flux, Duration windowDuration, int windowMaxSize) {
```

Comment:
could use ReactorUtils::throttle defined on https://github.com/linagora/james-project/pull/3395

Reply:
Could be done in a follow up refacto, lets not stack dependencies